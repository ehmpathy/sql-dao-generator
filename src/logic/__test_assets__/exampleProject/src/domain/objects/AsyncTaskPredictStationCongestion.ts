import { DomainEntity } from 'domain-objects';
import { AsyncTask, AsyncTaskStatus } from 'simple-async-tasks';

/**
 * an async task which predicts station congestion by comparing expected arrival time with estimated arrival time
 */
export interface AsyncTaskPredictStationCongestion extends AsyncTask {
  id?: number;
  uuid?: string;
  createdAt?: Date;
  updatedAt?: Date;
  status: AsyncTaskStatus;

  /**
   * the station to run this prediction for
   */
  stationUuid: string;

  /**
   * the event to run the prediction on
   *
   * note
   * - this has a nested dependency on a train, which we use to test sql-schema-control order
   */
  trainLocatedEventUuid: string;
}
export class AsyncTaskPredictStationCongestion
  extends DomainEntity<AsyncTaskPredictStationCongestion>
  implements AsyncTaskPredictStationCongestion
{
  public static unique = ['stationUuid', 'trainLocatedEventUuid'];
  public static updatable = ['status'];
}
