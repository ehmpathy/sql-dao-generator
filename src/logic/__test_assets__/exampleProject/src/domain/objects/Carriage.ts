import { DomainEntity } from 'domain-objects';

export enum CarriagePurpose {
  PASSENGER = 'PASSENGER',
  FREIGHT = 'FREIGHT',
}

/**
 * a carriage carries things or people in a train
 */
export interface Carriage {
  id?: number;
  uuid?: string;
  cin: string; // human readable identifier for carriage
  carries: CarriagePurpose;
  capacity: number; // for passenger carriages, number of passengers; for freight, the volume
}
export class Carriage extends DomainEntity<Carriage> implements Carriage {
  public static primary = ['uuid'] as const; // there is nothing that naturally identifies a carriage -> it is unique only on the unique identifier tracked in our db
  public static unique = ['uuid'] as const; // there is nothing that naturally identifies a carriage -> it is unique only on the unique identifier tracked in our db
  public static updatable = [];
}
