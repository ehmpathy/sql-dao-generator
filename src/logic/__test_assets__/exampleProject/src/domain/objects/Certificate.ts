import { DomainValueObject } from 'domain-objects';

export enum CertificateType {
  LOCOMOTIVE_DRIVING = 'LOCOMOTIVE_DRIVING',
  ENGINE_MAINTENANCE = 'ENGINE_MAINTENANCE',
  CARRIAGE_MAINTENANCE = 'CARRIAGE_MAINTENANCE',
}

export interface Certificate {
  id?: number;
  type: CertificateType;
  industryId: string;
}
export class Certificate
  extends DomainValueObject<Certificate>
  implements Certificate {}
