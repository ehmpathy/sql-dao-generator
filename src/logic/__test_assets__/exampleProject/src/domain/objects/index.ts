export * from './AsyncTaskPredictStationCongestion'
export * from './Carriage';
export * from './Certificate';
export * from './Engineer';
export * from './Geocode';
export * from './Train';
export * from './Locomotive';
export * from './TrainLocatedEvent';
export * from './Station';
export * from './Invoice';
export * from './InvoiceLineItem';
export * from './Price';
