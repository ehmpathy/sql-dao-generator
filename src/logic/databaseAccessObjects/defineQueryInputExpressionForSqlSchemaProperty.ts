import { camelCase } from 'change-case';
import { DomainObjectPropertyMetadata } from 'domain-objects-metadata';
import { SqlSchemaPropertyMetadata } from '../../domain/objects/SqlSchemaPropertyMetadata';
import { SqlSchemaReferenceMethod } from '../../domain/objects/SqlSchemaReferenceMetadata';
import { SqlSchemaToDomainObjectRelationship } from '../../domain/objects/SqlSchemaToDomainObjectRelationship';

export const defineQueryInputExpressionForSqlSchemaProperty = ({
  sqlSchemaProperty,
  domainObjectProperty,
  allSqlSchemaRelationships,
}: {
  sqlSchemaName: string;
  sqlSchemaProperty: SqlSchemaPropertyMetadata;
  domainObjectProperty: DomainObjectPropertyMetadata;
  allSqlSchemaRelationships: SqlSchemaToDomainObjectRelationship[];
}): string => {
  // simple case: non references
  if (!sqlSchemaProperty.isArray && !sqlSchemaProperty.reference) return `:${domainObjectProperty.name}`; // yesql notation

  // since sql-schema-generator does not support arrays of non-references, we can guarantee that its either a reference or an array of references now
  if (!sqlSchemaProperty.reference) throw new Error('unexpected code path, not supported. this is probably a bug'); // fail fast if our expectation is not met though

  // since we know its a reference, lookup the referenced sqlSchemaRelationship
  const referencedSqlSchemaRelationship = allSqlSchemaRelationships.find(
    (rel) => rel.name.domainObject === sqlSchemaProperty.reference!.of.name,
  );
  if (!referencedSqlSchemaRelationship)
    throw new Error('could not find referenced sql schema relationship. this is a bug within sql-dao-generator'); // fail fast, this should not occur
  const referencedSqlSchemaName = referencedSqlSchemaRelationship.name.sqlSchema;

  // handle solo reference
  if (!sqlSchemaProperty.isArray) {
    // if its a solo implicit uuid reference, then lookup the id by uuid as the input expression
    if (sqlSchemaProperty.reference.method === SqlSchemaReferenceMethod.IMPLICIT_BY_UUID)
      return `(SELECT id FROM ${referencedSqlSchemaName} WHERE ${referencedSqlSchemaName}.uuid = :${domainObjectProperty.name})`;

    // if its a nested reference, then it will have the id on it already
    if (sqlSchemaProperty.reference.method === SqlSchemaReferenceMethod.DIRECT_BY_NESTING)
      return `:${camelCase(sqlSchemaProperty.name)}`;
  }

  // handle array of references
  if (sqlSchemaProperty.isArray) {
    // if its a solo implicit uuid reference, then lookup the ids by uuids as the input expression
    if (sqlSchemaProperty.reference.method === SqlSchemaReferenceMethod.IMPLICIT_BY_UUID)
      return `
    (
      SELECT COALESCE(array_agg(${referencedSqlSchemaName}.id ORDER BY ${referencedSqlSchemaName}_ref.array_order_index), array[]::bigint[]) AS array_agg
      FROM ${referencedSqlSchemaName}
      JOIN unnest(:${domainObjectProperty.name}) WITH ORDINALITY
        AS ${referencedSqlSchemaName}_ref (uuid, array_order_index)
        ON ${referencedSqlSchemaName}.uuid = ${referencedSqlSchemaName}_ref.uuid
    )
      `.trim();

    // if its a nested reference, then it will be an array of ids arlready
    if (sqlSchemaProperty.reference.method === SqlSchemaReferenceMethod.DIRECT_BY_NESTING)
      return `:${camelCase(sqlSchemaProperty.name)}`;
  }

  // fail fast if we reach here, not expected
  throw new Error('unexpected code path. this a bug');
};
