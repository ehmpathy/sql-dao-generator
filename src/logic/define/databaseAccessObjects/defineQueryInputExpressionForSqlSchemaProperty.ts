import { camelCase, snakeCase } from 'change-case';
import { DomainObjectPropertyMetadata } from 'domain-objects-metadata';
import { SqlSchemaPropertyMetadata } from '../../../domain/objects/SqlSchemaPropertyMetadata';
import { SqlSchemaReferenceMethod } from '../../../domain/objects/SqlSchemaReferenceMetadata';
import { SqlSchemaToDomainObjectRelationship } from '../../../domain/objects/SqlSchemaToDomainObjectRelationship';
import { UnexpectedCodePathDetectedError } from '../../UnexpectedCodePathDetectedError';

export const defineQueryInputExpressionForSqlSchemaProperty = ({
  sqlSchemaName,
  sqlSchemaProperty,
  domainObjectProperty,
  allSqlSchemaRelationships,
}: {
  sqlSchemaName: string;
  sqlSchemaProperty: SqlSchemaPropertyMetadata;
  domainObjectProperty: DomainObjectPropertyMetadata;
  allSqlSchemaRelationships: SqlSchemaToDomainObjectRelationship[];
}): string => {
  if (!domainObjectProperty)
    throw new UnexpectedCodePathDetectedError({
      reason: 'no domain object property for sql schema property ',
      domainObjectName: camelCase(sqlSchemaName),
      domainObjectPropertyName: camelCase(sqlSchemaName),
    }); // fail fast if our expectation is not met though

  // simple case: non references
  if (!sqlSchemaProperty.reference) return `:${domainObjectProperty.name}`; // yesql notation

  // since we know its a reference, lookup the referenced sqlSchemaRelationship
  const referencedSqlSchemaRelationship = allSqlSchemaRelationships.find(
    (rel) => rel.name.domainObject === sqlSchemaProperty.reference!.of.name,
  );
  if (!referencedSqlSchemaRelationship)
    throw new UnexpectedCodePathDetectedError({
      reason: 'could not find referenced sql schema relationship for defining query input expression',
      domainObjectName: snakeCase(sqlSchemaName),
      domainObjectPropertyName: domainObjectProperty.name,
    }); // fail fast, this should not occur
  const referencedSqlSchemaName = referencedSqlSchemaRelationship.name.sqlSchema;

  // handle solo reference
  if (!sqlSchemaProperty.isArray) {
    // if its a solo implicit uuid reference, then lookup the id by uuid as the input expression
    if (sqlSchemaProperty.reference.method === SqlSchemaReferenceMethod.IMPLICIT_BY_UUID)
      return `(SELECT id FROM ${referencedSqlSchemaName} WHERE ${referencedSqlSchemaName}.uuid = :${domainObjectProperty.name})`;

    // if its a nested reference, then it will have the id on it already
    if (sqlSchemaProperty.reference.method === SqlSchemaReferenceMethod.DIRECT_BY_NESTING)
      return `:${camelCase(sqlSchemaProperty.name)}`;
  }

  // handle array of references
  if (sqlSchemaProperty.isArray) {
    // if its a solo implicit uuid reference, then lookup the ids by uuids as the input expression
    if (sqlSchemaProperty.reference.method === SqlSchemaReferenceMethod.IMPLICIT_BY_UUID)
      return `
    (
      SELECT COALESCE(array_agg(${referencedSqlSchemaName}.id ORDER BY ${referencedSqlSchemaName}_ref.array_order_index), array[]::bigint[]) AS array_agg
      FROM ${referencedSqlSchemaName}
      JOIN unnest(:${domainObjectProperty.name}::uuid[]) WITH ORDINALITY
        AS ${referencedSqlSchemaName}_ref (uuid, array_order_index)
        ON ${referencedSqlSchemaName}.uuid = ${referencedSqlSchemaName}_ref.uuid
    )
      `.trim();

    // if its a nested reference, then it will be an array of ids arlready
    if (sqlSchemaProperty.reference.method === SqlSchemaReferenceMethod.DIRECT_BY_NESTING)
      return `:${camelCase(sqlSchemaProperty.name)}`;
  }

  // fail fast if we reach here, not expected
  throw new UnexpectedCodePathDetectedError({
    reason: 'did not handle the request with any defined conditions, for query input expression',
    domainObjectName: camelCase(sqlSchemaName),
    domainObjectPropertyName: domainObjectProperty.name,
  }); // fail fast if our expectation is not met though
};
