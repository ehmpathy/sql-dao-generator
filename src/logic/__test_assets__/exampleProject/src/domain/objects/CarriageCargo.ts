import { DomainEntity, Ref } from 'domain-objects';
import { Carriage } from './Carriage';

/**
 * .what = cargo which has been allocated to a carriage at a given point in time
 */
export interface CarriageCargo {
  id?: number;
  uuid?: string;

  /**
   * the itinerary during which it is allocated
   */
  itineraryUuid: string; // todo: ref to an itinerary once we declare it

  /**
   * the carriage it has been allocated to
   */
  carriageRef: Ref<typeof Carriage>;

  /**
   * the slot that it has been allocated to on the carriage
   */
  slot: number;

  /**
   * the cargo that has been allocated
   */
  cargoExid: null | string;
}
export class CarriageCargo extends DomainEntity<CarriageCargo> implements CarriageCargo {
  public static primary = ['uuid'] as const;
  public static unique = ['itineraryUuid', 'carriageRef', 'slot'] as const;
  public static updatable = ['cargoExid'];
}
