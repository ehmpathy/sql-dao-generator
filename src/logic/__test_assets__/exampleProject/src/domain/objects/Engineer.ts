import { DomainEntity } from 'domain-objects';
import { Certificate } from './Certificate';

/**
 * a train engineer drives and maintains the train
 */
export interface TrainEngineer {
  id?: number;
  uuid?: string;
  socialSecurityNumberHash: string; // the hash of their social security number
  certificates: Certificate[]; // an array of what certificates they've earned
  name: string;
}
export class TrainEngineer extends DomainEntity<TrainEngineer> implements TrainEngineer {
  public static unique = ['socialSecurityNumberHash'];
  public static updatable = ['name', 'certificates'];
}
