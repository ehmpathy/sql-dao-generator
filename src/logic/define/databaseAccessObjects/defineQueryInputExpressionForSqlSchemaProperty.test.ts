import {
  DomainObjectPropertyType,
  DomainObjectVariant,
} from 'domain-objects-metadata';

import { SqlSchemaReferenceMethod } from '../../../domain/objects/SqlSchemaReferenceMetadata';
import { SqlSchemaToDomainObjectRelationship } from '../../../domain/objects/SqlSchemaToDomainObjectRelationship';
import { defineQueryInputExpressionForSqlSchemaProperty } from './defineQueryInputExpressionForSqlSchemaProperty';

describe('defineQueryInputExpressionForSqlSchemaProperty', () => {
  it('should define the input expression correctly for a standard column-lookup property', () => {
    const expression = defineQueryInputExpressionForSqlSchemaProperty({
      sqlSchemaName: 'train_engineer',
      sqlSchemaProperty: {
        name: 'social_security_number_hash',
        isArray: false,
        isNullable: false,
        isUpdatable: false,
        isDatabaseGenerated: false,
        reference: null,
      },
      domainObjectProperty: {
        name: 'socialSecurityNumberHash',
        type: DomainObjectPropertyType.STRING,
      },
      allSqlSchemaRelationships: [], // not needed for this one
    });
    expect(expression).toEqual(':socialSecurityNumberHash');
  });
  it('should define the input expression correctly for a solo IMPLICIT_BY_UUID reference', () => {
    const expression = defineQueryInputExpressionForSqlSchemaProperty({
      sqlSchemaName: 'train',
      sqlSchemaProperty: {
        name: 'lead_engineer_id',
        isArray: false,
        isNullable: false,
        isUpdatable: false,
        isDatabaseGenerated: false,
        reference: {
          method: SqlSchemaReferenceMethod.IMPLICIT_BY_UUID,
          of: {
            name: 'TrainEngineer',
            extends: DomainObjectVariant.DOMAIN_ENTITY,
          },
        },
      },
      domainObjectProperty: {
        name: 'leadEngineerUuid',
        type: DomainObjectPropertyType.STRING,
      },
      allSqlSchemaRelationships: [
        new SqlSchemaToDomainObjectRelationship({
          name: { domainObject: 'TrainEngineer', sqlSchema: 'train_engineer' },
          properties: [],
          decorations: { unique: { sqlSchema: null, domainObject: null } },
        }),
      ], // not needed for this one
    });
    expect(expression).toEqual(
      '(SELECT id FROM train_engineer WHERE train_engineer.uuid = :leadEngineerUuid)',
    );
  });
  it('should define the input expression correctly for a solo DIRECT_BY_NESTING reference', () => {
    const expression = defineQueryInputExpressionForSqlSchemaProperty({
      sqlSchemaName: 'train_located_event',
      sqlSchemaProperty: {
        name: 'geocode_id',
        isArray: false,
        isNullable: false,
        isUpdatable: false,
        isDatabaseGenerated: false,
        reference: {
          method: SqlSchemaReferenceMethod.DIRECT_BY_NESTING,
          of: {
            name: 'Geocode',
            extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT,
          },
        },
      },
      domainObjectProperty: {
        name: 'geocode',
        type: DomainObjectPropertyType.REFERENCE,
        of: {
          name: 'Geocode',
          extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT,
        },
      },
      allSqlSchemaRelationships: [
        new SqlSchemaToDomainObjectRelationship({
          name: { domainObject: 'Geocode', sqlSchema: 'geocode' },
          properties: [
            {
              domainObject: {
                name: 'latitude',
                type: DomainObjectPropertyType.NUMBER,
              },
              sqlSchema: {
                name: 'latitude',
                isArray: false,
                isNullable: false,
                isUpdatable: false,
                isDatabaseGenerated: false,
                reference: null,
              },
            },
            {
              domainObject: {
                name: 'longitude',
                type: DomainObjectPropertyType.NUMBER,
              },
              sqlSchema: {
                name: 'longitude',
                isArray: false,
                isNullable: false,
                isUpdatable: false,
                isDatabaseGenerated: false,
                reference: null,
              },
            },
          ],
          decorations: { unique: { sqlSchema: null, domainObject: null } },
        }),
      ],
    });
    expect(expression).toEqual(':geocodeId');
  });
  it('should define the input expression correctly for an array of IMPLICIT_BY_UUID references', () => {
    const expression = defineQueryInputExpressionForSqlSchemaProperty({
      sqlSchemaName: 'train',
      sqlSchemaProperty: {
        name: 'assigned_engineer_ids',
        isArray: true,
        isNullable: false,
        isUpdatable: false,
        isDatabaseGenerated: false,
        reference: {
          method: SqlSchemaReferenceMethod.IMPLICIT_BY_UUID,
          of: {
            name: 'TrainEngineer',
            extends: DomainObjectVariant.DOMAIN_ENTITY,
          },
        },
      },
      domainObjectProperty: {
        name: 'assignedEngineerUuids',
        type: DomainObjectPropertyType.ARRAY,
        of: {
          type: DomainObjectPropertyType.STRING,
        },
      },
      allSqlSchemaRelationships: [
        new SqlSchemaToDomainObjectRelationship({
          name: { domainObject: 'TrainEngineer', sqlSchema: 'train_engineer' },
          properties: [],
          decorations: { unique: { sqlSchema: null, domainObject: null } },
        }),
      ], // not needed for this one
    });
    expect(expression).toContain('COALESCE(array_agg(train_engineer.id'); // should select an array of ids, with coalesce to treat empty array
    expect(expression).toContain(
      'array_agg(train_engineer.id ORDER BY train_engineer_ref.array_order_index)',
    ); // should select them by persisted array order
    expect(expression).toContain('FROM train_engineer'); // from the right table
    expect(expression).toContain(
      'JOIN unnest(:assignedEngineerUuids::uuid[]) WITH ORDINALITY',
    ); // using 'unnest' for performance gains,  using "ordinality" so that we can preserve persisted array_order_index
    expect(expression).toContain(
      'train_engineer_ref (uuid, array_order_index)',
    ); // name the unnested w/ ordinality table well
    expect(expression).toContain(
      'ON train_engineer.uuid = train_engineer_ref.uuid',
    ); // filtered on the right id
    expect(expression).toMatchSnapshot();
  });
  it('should define the input expression correctly for an array of DIRECT_BY_NESTING reference', () => {
    const expression = defineQueryInputExpressionForSqlSchemaProperty({
      sqlSchemaName: 'train_located_event',
      sqlSchemaProperty: {
        name: 'geocode_ids',
        isArray: true,
        isNullable: false,
        isUpdatable: false,
        isDatabaseGenerated: false,
        reference: {
          method: SqlSchemaReferenceMethod.DIRECT_BY_NESTING,
          of: {
            name: 'Geocode',
            extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT,
          },
        },
      },
      domainObjectProperty: {
        name: 'geocodes',
        type: DomainObjectPropertyType.ARRAY,
        of: {
          type: DomainObjectPropertyType.REFERENCE,
          of: {
            name: 'Geocode',
            extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT,
          },
        },
      },
      allSqlSchemaRelationships: [
        new SqlSchemaToDomainObjectRelationship({
          name: { domainObject: 'Geocode', sqlSchema: 'geocode' },
          properties: [
            {
              domainObject: {
                name: 'latitude',
                type: DomainObjectPropertyType.NUMBER,
              },
              sqlSchema: {
                name: 'latitude',
                isArray: false,
                isNullable: false,
                isUpdatable: false,
                isDatabaseGenerated: false,
                reference: null,
              },
            },
            {
              domainObject: {
                name: 'longitude',
                type: DomainObjectPropertyType.NUMBER,
              },
              sqlSchema: {
                name: 'longitude',
                isArray: false,
                isNullable: false,
                isUpdatable: false,
                isDatabaseGenerated: false,
                reference: null,
              },
            },
          ],
          decorations: { unique: { sqlSchema: null, domainObject: null } },
        }),
      ],
    });
    expect(expression).toEqual(':geocodeIds');
  });
});
