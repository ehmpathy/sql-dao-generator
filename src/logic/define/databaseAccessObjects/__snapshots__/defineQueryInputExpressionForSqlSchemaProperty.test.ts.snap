// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineQueryInputExpressionForSqlSchemaProperty should define the input expression correctly for an array of IMPLICIT_BY_UUID references 1`] = `
"(
      SELECT COALESCE(array_agg(train_engineer.id ORDER BY train_engineer_ref.array_order_index), array[]::bigint[]) AS array_agg
      FROM train_engineer
      JOIN unnest(:assignedEngineerUuids::uuid[]) WITH ORDINALITY
        AS train_engineer_ref (uuid, array_order_index)
        ON train_engineer.uuid = train_engineer_ref.uuid
    )"
`;
